src/drivers/ata-pci.c

```diff
@@ -66,36 +134,209 @@ static const char *ata_pcivendor2str(pci_dev d)
+                        ata_pci, apci, boolean, write,
+                        void *, buf, range, blocks, status_handler, sh)
+{
+    ata_pci apci = bound(apci);
+    ata_pci_req req = allocate(apci->h, sizeof(*req));
+    if (req == INVALID_ADDRESS) {
+        apply_status(timm("result", "request allocation failed"), sh);
+        return;
+    }
+    req->write = bound(write);
+    req->buf = buf;
+    req->remain = blocks;
+    req->sh = sh;
+    req->s = STATUS_OK;
+    u64 irqflags = spin_lock_irq(&apci->lock);
+    boolean idle = list_empty(&apci->reqs);
+    list_push_back(&apci->reqs, &req->l);
+    spin_unlock_irq(&apci->lock, irqflags);
+    if (idle)
+        ata_pci_service_reqs(apci);
+}
+
+define_closure_function(1, 0, void, ata_pci_irq,
+                        ata_pci, apci)
+{
+    ata_pci apci = bound(apci);
+    pci_bar_write_1(&apci->bmr, ATA_BMR_CMD(ATA_PRIMARY), 0);
+    u8 status = pci_bar_read_1(&apci->bmr, ATA_BMR_STATUS(ATA_PRIMARY));
+    boolean error = ata_clear_irq(apci->ata);
+    ata_debug("%s: BMR status 0x%02x, ATA %s\n", __func__, status,
+              error ? "error" : "OK");
+    if (status & ATA_BMR_STATUS_ERR) {
+        /* Clear error status */
+        pci_bar_write_1(&apci->bmr, ATA_BMR_STATUS(ATA_PRIMARY),
+            ATA_BMR_STATUS_ERR);
+
+        error = true;
+    }
+    u64 irqflags = spin_lock_irq(&apci->lock);
+    if (!list_empty(&apci->reqs)) {
+        if (error) {
+            ata_pci_req req = struct_from_list(list_begin(&apci->reqs),
+                ata_pci_req, l);
+            req->s = timm("result", "I/O error");
+        }
+        enqueue(bhqueue, &apci->service);
+    }
+    spin_unlock_irq(&apci->lock, irqflags);
+}
+
+define_closure_function(1, 0, void, ata_pci_service,
+                        ata_pci, apci)
+{
+    ata_pci_service_reqs(bound(apci));
+}
+
+closure_function(3, 1, boolean, ata_pci_probe,
+                 heap, general, heap, contiguous, storage_attach, a,
                  pci_dev, d)
 {
     heap general = bound(general);
+    heap contiguous = bound(contiguous);
+    ata_debug("%s: class %d subclass %d\n", __func__, pci_get_class(d), pci_get_subclass(d));
     if (pci_get_class(d) != PCIC_STORAGE || pci_get_subclass(d) != PCIS_STORAGE_IDE)
         return false;
 
-    struct ata *dev = ata_pci_alloc(general, d);
-    if (!ata_probe(dev)) {
-        ata_dealloc(dev);
+    ata_pci dev = ata_pci_alloc(general, contiguous, d);
+    if (!ata_probe(dev->ata)) {
+        ata_pci_dealloc(general, contiguous, dev);
         return false;
     }
 
+    /* Align the PRDT to ensure it doesn't cross a 64K boundary. */
+    dev->prdt_phys = physical_from_virtual(dev->prdt);
+    u64 boundary = pad(dev->prdt_phys, 64 * KB);
+    int align_offset =
+            (dev->prdt_phys + PRDT_ENTRIES * sizeof(struct prd) > boundary) ?
+                    boundary - dev->prdt_phys : 0;
+    dev->prdt_phys += align_offset;
+    assert(dev->prdt_phys < U64_FROM_BIT(32));
```

Comment:
There is no explicit guarantee, but I verified that when launching AWS instances with more than 4GB, memory allocation starts from low addresses that fit in 32 bits, and since this code is executed at startup we should be safe.